src/pandas/types/numeric.h

```diff
@@ -44,14 +106,153 @@ class PANDAS_EXPORT IntegerArray : public NumericArray<TYPE> {
 
   bool owns_data() const override;
 
+  Status EnsureMutable();
+
+  // In the case where there are any nulls in this array,
+  // this function will be faster than calling GetNullCount() > 0.
+  bool HasNulls() const {
+    for (auto ii = 0; ii < this->length_; ++ii) {
+      if (IsNull(ii)) { return true; }
+    }
+    return false;
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetNull(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::SetBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetValid(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::ClearBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Returns whether or not the value at the specified index has
+  // been marked as null.
+  bool IsNull(int64_t i) const {
+    if (valid_bits_) {
+      // TODO: range check?
+      return BitUtil::GetBit(valid_bits_->data(), this->offset_ + i);
+    } else {
+      return false;
+    }
+  }
+
+  // Copy the list of nulls (aka valid bits) to a buffer
+  Status CopyNulls(int64_t length, std::shared_ptr<Buffer>* out) const {
+    // Pre-condition is that we already have valid_bits_ set
+    return valid_bits_->Copy(this->offset_, BitUtil::BytesForBits(length), out);
```

Comment:
Offset needs to be used as a bit offset. Can use CopyBitmap here (which hasn't been properly implemented yet, but the signature is okay)